File: android/src/main/kotlin/com/loopit/minis/imgedit/LayerStack.kt
package com.loopit.minis.imgedit

/**
 * Mirrors the layer model from improvement2.md.
 *
 * The stack is ordered bottom-to-top; render code in `cpp/imgedit/pipeline.cpp`
 * walks it in order and composites each layer into the offscreen FBO.
 *
 * This Kotlin side stores layer params verbatim and forwards them to JNI on
 * every change — the C++ renderer is the source of truth for pixels.
 */
class LayerStack {

    enum class Kind { BASE_IMAGE, ADJUSTMENT, FILTER, STICKER, TEXT, DRAW, MASK, EMOJI }

    data class Layer(
        val id: Int,
        val kind: Kind,
        var params: Map<String, Any?>,
        var visible: Boolean = true,
        var opacity: Double = 1.0,
        var blend: String = "normal",
    )

    private val items = ArrayList<Layer>()
    private val adjust = HashMap<String, Double>()
    private var lutPath: String = ""
    private var lutIntensity: Double = 0.0
    private var crop: Map<String, Any?> = emptyMap()
    private var canvasWidth: Int = 0
    private var canvasHeight: Int = 0

    fun reset(w: Int, h: Int) {
        canvasWidth = w
        canvasHeight = h
        items.clear()
        adjust.clear()
        lutPath = ""
        lutIntensity = 0.0
        crop = emptyMap()
        items += Layer(ImageEditEngine.nextLayerId(), Kind.BASE_IMAGE, emptyMap())
    }

    fun push(typeWire: String, params: Map<String, Any?>): Int {
        val kind = parseKind(typeWire)
        val layer = Layer(ImageEditEngine.nextLayerId(), kind, params)
        items += layer
        return layer.id
    }

    fun pushMask(maskHandle: Map<String, Any?>): Int {
        val layer = Layer(ImageEditEngine.nextLayerId(), Kind.MASK, maskHandle)
        items += layer
        return layer.id
    }

    fun update(id: Int, params: Map<String, Any?>): Map<String, Any?> {
        val layer = items.firstOrNull { it.id == id } ?: return emptyMap()
        val prev = layer.params
        layer.params = params
        return prev
    }

    fun remove(id: Int): Layer? {
        val idx = items.indexOfFirst { it.id == id }
        if (idx < 0) return null
        return items.removeAt(idx)
    }

    fun reorder(id: Int, target: Int): Int {
        val from = items.indexOfFirst { it.id == id }
        if (from < 0) return -1
        val layer = items.removeAt(from)
        val clamped = target.coerceIn(0, items.size)
        items.add(clamped, layer)
        return from
    }

    fun applyAdjust(key: String, value: Double) {
        adjust[key] = value
    }

    fun applyFilter(lutPath: String, intensity: Double) {
        this.lutPath = lutPath
        this.lutIntensity = intensity
    }

    fun applyCrop(rect: Map<String, Any?>, rotationDeg: Double, persp: List<Double>?) {
        crop = mapOf(
            "rect" to rect,
            "rotationDeg" to rotationDeg,
            "persp" to (persp ?: emptyList<Double>()),
        )
    }

    fun currentAdjust(key: String): Double = adjust[key] ?: 0.0
    fun currentLut(): String = lutPath
    fun currentLutIntensity(): Double = lutIntensity
    fun currentCrop(): Map<String, Any?> = crop

    fun snapshot(): Snapshot = Snapshot(
        layers = items.map { it.copy() },
        adjust = HashMap(adjust),
        lutPath = lutPath,
        lutIntensity = lutIntensity,
        crop = crop,
    )

    fun restore(snapshot: Snapshot) {
        items.clear()
        items.addAll(snapshot.layers)
        adjust.clear()
        adjust.putAll(snapshot.adjust)
        lutPath = snapshot.lutPath
        lutIntensity = snapshot.lutIntensity
        crop = snapshot.crop
    }

    data class Snapshot(
        val layers: List<Layer>,
        val adjust: Map<String, Double>,
        val lutPath: String,
        val lutIntensity: Double,
        val crop: Map<String, Any?>,
    )

    private fun parseKind(wire: String): Kind = when (wire) {
        "baseImage" -> Kind.BASE_IMAGE
        "adjustment" -> Kind.ADJUSTMENT
        "filter" -> Kind.FILTER
        "sticker" -> Kind.STICKER
        "text" -> Kind.TEXT
        "draw" -> Kind.DRAW
        "mask" -> Kind.MASK
        "emoji" -> Kind.EMOJI
        else -> Kind.ADJUSTMENT
    }
}

File: ios/Classes/ImgEdit/LayerStack.swift
import Foundation

/// Mirrors `LayerStack.kt` and the layer model in improvement2.md.
public final class LayerStack {

    public enum Kind: String {
        case baseImage
        case adjustment
        case filter
        case sticker
        case text
        case draw
        case mask
        case emoji
    }

    public struct Layer {
        public let id: Int
        public let kind: Kind
        public var params: [String: Any]
        public var visible: Bool
        public var opacity: Double
        public var blend: String
    }

    private static var nextId: Int = 1
    private(set) var items: [Layer] = []
    private var adjust: [String: Double] = [:]
    private(set) var lutPath: String = ""
    private(set) var lutIntensity: Double = 0
    private(set) var crop: [String: Any] = [:]
    private(set) var width: Int = 0
    private(set) var height: Int = 0

    public func reset(width: Int, height: Int) {
        self.width = width
        self.height = height
        items.removeAll()
        adjust.removeAll()
        lutPath = ""
        lutIntensity = 0
        crop = [:]
        items.append(Layer(
            id: nextLayerId(),
            kind: .baseImage,
            params: [:],
            visible: true,
            opacity: 1,
            blend: "normal"
        ))
    }

    @discardableResult
    public func push(type: String, params: [String: Any]) -> Int {
        let kind = parseKind(type)
        let layer = Layer(
            id: nextLayerId(),
            kind: kind,
            params: params,
            visible: true,
            opacity: 1,
            blend: "normal"
        )
        items.append(layer)
        return layer.id
    }

    @discardableResult
    public func pushMask(_ payload: [String: Any]) -> Int {
        let layer = Layer(
            id: nextLayerId(),
            kind: .mask,
            params: payload,
            visible: true,
            opacity: 1,
            blend: "normal"
        )
        items.append(layer)
        return layer.id
    }

    @discardableResult
    public func update(id: Int, params: [String: Any]) -> [String: Any] {
        guard let idx = items.firstIndex(where: { $0.id == id }) else { return [:] }
        let prev = items[idx].params
        items[idx].params = params
        return prev
    }

    @discardableResult
    public func remove(id: Int) -> Layer? {
        guard let idx = items.firstIndex(where: { $0.id == id }) else { return nil }
        return items.remove(at: idx)
    }

    @discardableResult
    public func reorder(id: Int, target: Int) -> Int {
        guard let from = items.firstIndex(where: { $0.id == id }) else { return -1 }
        let layer = items.remove(at: from)
        let clamped = max(0, min(items.count, target))
        items.insert(layer, at: clamped)
        return from
    }

    public func applyAdjust(key: String, value: Double) {
        adjust[key] = value
    }

    public func applyFilter(lutPath: String, intensity: Double) {
        self.lutPath = lutPath
        self.lutIntensity = intensity
    }

    public func applyCrop(rect: [String: Any], rotationDeg: Double, persp: [Double]?) {
        crop = [
            "rect": rect,
            "rotationDeg": rotationDeg,
            "persp": persp ?? [],
        ]
    }

    public func adjustValue(forKey key: String) -> Double {
        return adjust[key] ?? 0
    }

    public func currentLut() -> String { lutPath }
    public func currentLutIntensity() -> Double { lutIntensity }
    public func currentCrop() -> [String: Any] { crop }

    public func snapshot() -> Snapshot {
        Snapshot(
            layers: items,
            adjust: adjust,
            lutPath: lutPath,
            lutIntensity: lutIntensity,
            crop: crop
        )
    }

    public func restore(_ snapshot: Snapshot) {
        items = snapshot.layers
        adjust = snapshot.adjust
        lutPath = snapshot.lutPath
        lutIntensity = snapshot.lutIntensity
        crop = snapshot.crop
    }

    public struct Snapshot {
        public let layers: [Layer]
        public let adjust: [String: Double]
        public let lutPath: String
        public let lutIntensity: Double
        public let crop: [String: Any]
    }

    private func nextLayerId() -> Int {
        LayerStack.nextId += 1
        return LayerStack.nextId
    }

    private func parseKind(_ wire: String) -> Kind {
        return Kind(rawValue: wire) ?? .adjustment
    }
}
